src/ripple/rpc/handlers/Catalogue.cpp

```diff
@@ -0,0 +1,1139 @@
+            return rpcError(rpcLEDGER_MISSING);
+
+        // Process with diff against previous ledger
+        if (!outputLedger(currLedger, prevLedger->stateMap()))
+            return rpcError(
+                rpcINTERNAL, "Error occurred while processing ledgers");
+
+        ledgers_written++;
+
+        // Periodically flush the stream to prevent excessive memory usage in
+        // compression buffers Flush every 1000 ledgers as a reasonable balance
+        // between compression efficiency and memory usage
+        if (ledgers_written % 1000 == 0)
+        {
+            JLOG(context.j.info())
+                << "Flushing compression stream at ledger " << ledger_seq;
+            compStream->flush();
+        }
+
+        // Cycle the ledgers: current becomes previous, we'll load a new current
+        // next iteration
+        prevLedger = currLedger;
+    }
+
+    // flush and finish
+    compStream->flush();
+    compStream->reset();
+    outfile.flush();
+    outfile.close();
+
+    // Clear ledger references to release memory
+    prevLedger = nullptr;
+    currLedger = nullptr;
+
+    // Get the file size and update it in the header
+    if (stat(filepath.c_str(), &st) != 0)
+    {
+        JLOG(context.j.warn())
+            << "Could not get file size: " << std::strerror(errno);
+        return rpcError(
+            rpcINTERNAL, "failed to get file size for header update");
+    }
+
+    file_size = st.st_size;
+
+    // Update header with filesize
+    JLOG(context.j.info()) << "Updating file size in header: "
+                           << std::to_string(file_size) << " bytes";
+
+    header.filesize = file_size;
+    std::fstream updateFileSizeFile(
+        filepath.c_str(), std::ios::in | std::ios::out | std::ios::binary);
+    if (updateFileSizeFile.fail())
+        return rpcError(
+            rpcINTERNAL,
+            "cannot open file for updating filesize: " +
+                std::string(strerror(errno)));
+
+    updateFileSizeFile.seekp(0, std::ios::beg);
+    updateFileSizeFile.write(
+        reinterpret_cast<const char*>(&header), sizeof(CATLHeader));
+    updateFileSizeFile.close();
+
+    // Now compute the hash over the entire file
+    JLOG(context.j.info()) << "Computing catalogue hash...";
+
+    std::ifstream hashFile(filepath.c_str(), std::ios::in | std::ios::binary);
+    if (hashFile.fail())
+        return rpcError(
+            rpcINTERNAL,
+            "cannot open file for hashing: " + std::string(strerror(errno)));
+
+    // Initialize hasher
+    sha512_hasher hasher;
+
+    // Create a buffer for reading
+    std::vector<char> buffer(64 * 1024);  // 64K buffer
+
+    // Read and process the header portion
+    hashFile.read(buffer.data(), sizeof(CATLHeader));
+    if (hashFile.gcount() != sizeof(CATLHeader))
+        return rpcError(rpcINTERNAL, "failed to read header for hashing");
+
+    // Zero out the hash portion in the buffer for hash calculation
+    std::fill(
```

Comment:
This is kind of "belts and suspenders" mode, I guess, because the hash is already zeroed and the file only just written:

```c++
struct CATLHeader
{
    uint32_t magic = CATL;
    uint32_t min_ledger;
    uint32_t max_ledger;
    uint16_t version;
    uint16_t network_id;
    uint64_t filesize = 0;  // Total size of the file including header
    std::array<uint8_t, 64> hash = {};  // SHA-512 hash, initially set to zeros
};
```